### ISSUE TITLE
Document Process For Becoming a Committer

### ISSUE THREAD

**Issue by jvasile** [archesproject/arches#3321]
Arches has not established and documented the process by which a contributor becomes a committer.  It should establish that process and document it in the repo.  No contributor should be granted commit access who hasn't completed that process.  Committer status should be separate from employment status, which is just to say that folks who are hired at Farallon Geographics or Getty should have to travel the same path as anybody else.

#### Definition of Done
A process agreed to and documented in the repo.

#### Process for Resolution
Convene a conversation on the mailing list and go from there.

**Comment by jvasile** [comment]
@kfogel This is relevant to your interests.

**Comment by mradamcox** [comment]
@jvasile could you please comment on the existing documentation that we have here https://github.com/archesproject/arches/wiki as it relates to this ticket?

**Comment by jvasile** [comment]
@mradamcox, it was @kfogel who mentioned to me that this documentation exists, so either I will get that information from him or he will chime in here.  Either way, it's on my plate to cause one of those to occur.

**Comment by apeters** [comment]
@robgaston Check this out (https://github.com/v8/v8/wiki/Becoming-a-committer) for a good example of "How to become a committer".  I think we could follow something like this (with small tweaks).

**Comment by robgaston** [comment]
@apeters I agree that v8/Chromium provides a good approach for us here.  We may want to adjust some of the requirements (eg, the 3 reviewers requirement may be a lot given the relative size of our team/project), but this seems like a sound way forward to me.

**Comment by mradamcox** [comment]
@robgaston @apeters @jvasile @kfogel How was this decided and when did the discussion take place? One question I have: why is this new document completely separate from what was in discussion here https://github.com/archesproject/arches/issues/3320? Wouldn't it make sense to have Contribution guidelines and "becoming a committer" guidelines in the same place?

**Comment by mradamcox** [comment]
Thanks @apeters for moving this forward by drafting that document. I have a few small comments:

1. It seems to reference a hyperlink that doesn't exist: "...A committer is a contributor to the Arches projects' success and a citizen helping the project succeed. See Committer's Responsibility." There is no section titled "Committer's Responsibility" or link here. Should this reference be removed? In the v8 document "Committer's Responsibility" links to their general commit guidelines. We have guidelines like this, though they are still in multiple places: https://github.com/archesproject/arches/wiki/Contributing and https://github.com/archesproject/arches/wiki#contributing-to-arches. If this link is to be updated correctly, we should probably fix these disparate sources beforehand.

2. The document is named "Becoming a Contributor" though it is actually about becoming a committer. recommend changing name to "Becoming a Committer".

**Comment by kfogel** [comment]
@mradamcox FWIW I think it would be fine to just have "Becoming a Committer" as a section in `CONTRIBUTING.md` (as per your reference to #3320).

**Comment by mradamcox** [comment]
@kfogel I do think that would be cleaner, but we'll keep it separate for right now. The documents are now out for public comment, https://groups.google.com/forum/#!topic/archesproject/pIFb8pMI3qE, so it's possible that by the end of that period (about a month from now) we will have decided to combine them.

**Comment by kfogel** [comment]
A couple of comments about the become-a-committer process:

1. This is a relatively high-overhead process, relative to most projects (as far as I'm aware).  It asks people to self-nominate and to write "an explanation of why you should be a committer", which unintentionally penalizes those who are shy, lack self-confidence, or who don't have much prior open-source contribution experience and thus have less of a feel for where they stand with respect to other contributors and to the projects.

     Consider adding a "we might invite you out of the blue -- please say yes when we do!" clause, and a process for deciding how to issue such invitations.  (I can provide examples if you need; https://producingoss.com/en/committers.html also discusses this.)  Relying on self-nomination needlessly limits your committer pool.

2. The current guidelines imply that one must stay active in order to remain a committer (though they don't say precisely how active).  Is that really necessary?  Another way to think of commit access is that it's a statement about trusting someone's judgement because she's demonstrated that judgement by repeatedly submitting high-quality contributions. You trust her to know what she knows and what she doesn't know.  Maybe someone submits a lot of typo fixes, but never touches any really complex code.  But that's okay: her typo fixes were all correct and unambiguous improvement to the code.  Giving her commit access at this point will just save everyone time; she can still submit PRs for review if she starts getting into larger code changes (which she should of course be encouraged to do).

     I have commit access to GNU Emacs, for example, but there are large swaths of the code in which I'm not an expert and I would never dare to approve contributions for those areas (I might review others' contributions, but I would mention in my review that I'm not a domain expert for that part of the code).  When I want to make a change to those areas, I post the change for review first too.  Just because I *could* commit such a change directly doesn't mean I do -- I exercise judgement and seek help when I need it.

     Back to the question of expiration of commit access: if Arches trusted her judgement this year, why wouldn't Arches still trust it five years later, even if she didn't make many commits in the meantime?  Is there any reason why commit access should expire if the committer has not explicitly requested to be removed or listed as dormant?

**Comment by apeters** [comment]
Just responding to Karl's observations concerning this doc
https://github.com/archesproject/arches/wiki/Becoming-a-Committer:

1.  If you re-read this, you'll notice that the potential committer doesn't
self nominate, a current committer nominates the person and provides
information about them to the team.

Here's the text from the doc.

A current committer nominates you by slacking the team on the Archesproject
> development channel
>
>    - your first and last name
>
>
>    - a link to your Git profile
>
>
>    - an explanation of why you should be a committer,
>
>
>    - embedded list of links to revisions (about top 3) containing your
>    patches
>
>

2.  Looking at this document:
https://github.com/archesproject/arches/wiki/Becoming-a-Committer#maintaining-committer-status

There is no requirement for maintaining committer status.
We literally say,  "You don't really need to do much to maintain committer
status: just keep being awesome and helping the Arches project!"


Cheers,
Alexei

Director of Web Development - Farallon Geographics, Inc. - 971.227.3173

**Comment by kfogel** [comment]
If the process is to "ask someone to nominate you", that's essentially self-nomination.  My point is not about terminology, it's about where the project expects the initial action to come from.  If the project were to actively *invite* people who otherwise might not have thought to nominate themselves, then the project won't lose out on potential committers who are too shy to nominate themselves, or who have a different (and higher) personal standard for when they think they are ready to become a committer than the project itself has.

People can still self-nominate (that is, explicitly request that the process begin) if they want to.  I'm just saying that the project shouldn't depend only on that.  It should also be constantly considering the current contributors and inviting people to become committers when appropriate.

Regarding maintenance of committer status: what I meant was that the "keep ... helping the Arches project" doesn't actually need to be a criterion.  A person can vanish for two years, not participating at all, but when they come back and are ready to commit something, well if you trusted their judgement before there's no reason not to continue trusting it now.

In other words, if you really mean it about "there is no requirement for maintaining committer status", then remove the requirement that the person "keep ... helping the Arches project", because as currnetly phrased that's a requirement for maintaining committer status :-).
